docs/src/lucet-runtime/killswitch.md

```diff
@@ -367,11 +367,66 @@ handlers used with Lucet is that they may not lock on `KillState`'s
 
 As a consequence, a `KillSwitch` may fire during the handling of a guest fault.
 `sigaction` must mask `SIGALRM` so that a signal fired before the handler exits
-is discarded. If the signal behavior is to continue without effect, leave
-termination in place and continue to the guest. Otherwise the signal handler
-has determined it must return to the host, and it disables termination on the
-instance to avoid sending an erroneous SIGALRM immediately after swapping to
-the host context.
+does not preempt the handler. If the signal behavior is to continue without
+effect, leave termination in place and continue to the guest. A pending SIGALRM
+will be raised at this point and the instance will exit. Otherwise, the signal
+handler has determined it must return to the host, and must be sensitive to a
+possible in-flight `KillSwitch`..
+
+#### Instance-stopping guest fault with concurrent `KillSwitch`
+
+In the case we know three things:
+* The signal handler must return to host code
+* A `KillSwitch` may fire and deliver a `SIGALRM` at any point
+* A `SIGALRM` may already have been fired, pending on the handler returning
+
+First, we handle the risk of a `KillSwitch` firing: disable termination. If we
+acquire `terminable`, we know this is to the exclusion of any `KillSwitch`, and
+are safe to return. Otherwise, some `KillSwitch` has terminated, or is in the
+process of terminating, this guest's execution. This means a `SIGALRM` may be
+pending or imminent!
+
+A slightly simpler model is to consider that a `SIGALRM` may arrive in the
+future. For correctness, then, we only have to make sure we handle the signal
+we can be sent! We know that we must return to the host, and the guest fault
+that occurred is certainly more interesting than the guest termination, so we
+would like to preserve that information. There is no information or effect we
+want from the signal, so silence the alarm on `KillState`. This way, if we
+recieve the possible `SIGARLM`, we know to ignore it.
+
+An important question to ask is "How do we know the possible `SIGARLM` must be
+ignored? Could it not be for another instance later on that thread?" The answer
+is, in short, "We ensure it cannot!"
+
+The `SIGARLM` is thread-directed, so to be an alarm for some other reason,
+another instance would have to run and be terminated. To prevent this, we must
+prevent another instance from running. Additionally, if a `SIGALRM` is _in
+flight_, we need to stop and wait anyway. Since `KillSwitch` maintains a lock
+on `execution_domain` as long as it's attempting to terminate a guest, we can
+achieve both of these goals by taking, and immediately dropping, a lock on
+`execution_domain` before descheduling an instance.
+
+Even taking this lock is interesting:
+0. This lock could be taken while racing a `KillSwitch`, after it has observed it may fire but before advancing to take this lock.
```

Comment:
This is going to be another one of those moments where I ask a question without fully understanding what I am asking.  Good times!  I appreciate your patience with me.

I find myself dwelling on this sentence.  

Is it all possible to make it so that the KillSwitch can do the following atomically?
- observe that it may fire.
- take the lock. 

Probably not.  Which is why this case is here.  😊

Reply:
I think we're very close to a world where the answer is "yes", but just a bit short.

In a prior iteration of `KillSwitch`, we needed to be able to check "can the instance exit" in `context_asm.S`, which motivated the split at first. Nowdays that's all `{enter,exit}_guest_region`, without assembly-friendliness concerns.

The other weird constraint is that we can't take a lock in a signal handler. Since we would only do this in a SIGALRM case, and we SIGALRM a thread at most once, I think the real risk of this deadlocking is low, but I'd prefer to not play with fire (deadlocks) here. :grin: This is the only place we actually can't lock anymore, so a future refactor might be able to move this out to `deschedule` and join `execution_domain` and `terminable` into one piece of state.

Reply:
Thanks for explaining the historical context, why this separation was needed for assembly language concerns.  Joining `execution_domain` and `terminable` into a single piece of state sounds dreamy.  But that's not a dream to realize today.  😊  I appreciate you.  Thanks for your work.